doc/sphinx/language/core/records.rst

```diff
@@ -3,208 +3,313 @@
-in which case the correctness of :n:`@type__3` may rely on the instance :n:`@term__2` of :n:`@ident__2` and :n:`@term__2` may in turn depend on :n:`@ident__1`.
+   .. example:: Defining a record
+
+      The set of rational numbers may be defined as:
 
-.. example::
+      .. coqtop:: reset all
 
-   The set of rational numbers may be defined as:
+         Record Rat : Set := mkRat
+          { negative : bool
+          ; top : nat
+          ; bottom : nat
+          ; Rat_bottom_nonzero : 0 <> bottom
+          ; Rat_irreducible :
+              forall x y z:nat, (x * y) = top /\ (x * z) = bottom -> x = 1
+          }.
 
-   .. coqtop:: reset all
+      The :n:`Rat_*` fields depend on :n:`top` and :n:`bottom`.
+      :n:`Rat_bottom_nonzero` is a proof that :n:`bottom` (the denominator)
+      is not zero.  :n:`Rat_irreducible` is a proof that the fraction is in
+      lowest terms.
 
-      Record Rat : Set := mkRat
-       { sign : bool
-       ; top : nat
-       ; bottom : nat
-       ; Rat_bottom_cond : 0 <> bottom
-       ; Rat_irred_cond :
-           forall x y z:nat, (x * y) = top /\ (x * z) = bottom -> x = 1
-       }.
+.. _reuse_field_name:
 
-   Note here that the fields ``Rat_bottom_cond`` depends on the field ``bottom``
-   and ``Rat_irred_cond`` depends on both ``top`` and ``bottom``.
+   .. example:: Reusing a field name in multiple records
 
-Let us now see the work done by the :cmd:`Record` command. First the command
-generates a variant type definition with just one constructor:
-:n:`Variant @ident {* @binder } : @sort := @ident__0 {* @binder }`.
+      .. coqtop:: in
 
-To build an object of type :token:`ident`, provide the constructor
-:n:`@ident__0` with the appropriate number of terms filling the fields of the record.
+         Module A. Record R := { f : nat }. End A.
+         Module B. Record S := { f : nat }. End B.
 
-.. example::
+      .. coqtop:: all
 
-   Let us define the rational :math:`1/2`:
+         Check {| A.f := 0 |}.
+         Check {| B.f := 0 |}.
 
-    .. coqtop:: in
+.. _record_as_clause:
 
-       Theorem one_two_irred : forall x y z:nat, x * y = 1 /\ x * z = 2 -> x = 1.
-       Admitted.
+   .. example:: Using the "as" clause in a record definition
 
-       Definition half := mkRat true 1 2 (O_S 1) one_two_irred.
-       Check half.
+      .. coqtop:: all
 
-Alternatively, the following syntax allows creating objects by using named fields, as
-shown in this grammar. The fields do not have to be in any particular order, nor do they have
-to be all present if the missing ones can be inferred or prompted for
-(see :ref:`programs`).
+         Record MyRecord := { myfield : nat } as VarName.
+         About myfield. (* observe the MyRecord variable is named VarName *)
+         
+         (* This allows us to make VarName implicit for instance, without having to rename the variable as would be the case without the as clause. *)
```

Comment:
How does one rename the argument?

Reply:
If I have
```coq
Definition foo (x : nat) : nat := ... .
```
But I want the name of x to be different, I use:
```coq
Arguments foo y : rename.
```
If you don't put rename coq will complain.

Reply:
Hmm, if I use `Arguments bar HIMOM : rename`, what has to change in `Check (bar (HIMOM:={|bar:=0|}))`?

Reply:
@jfehrle That will stay the same. You can see the current names of the variables using the `About` command. So if you look at `About bar.` you will see that it says `Arguments bar HIMOM`, so writing exactly that, even with a rename will do nothing. If we later found that we wanted to change HIMOM to HIDAD for whatever reason, we could achieve this by writing `Arguments bar HIDAD : rename.` Notably, the syntax `bar (HIMOM:=_)` wouldn't work anymore and we would instead have to write `bar (HIDAD:=_)`.

Reply:
So with the "as HIMOM", you don't need ": rename" and vice-versa.  But both need an `Arguments` command.  Still a little puzzled why you use `bar (HIMOM:={|bar:=0|})` rather than just `bar {|bar:=0|}`.  Is it because the latter needs to be simplified to match the first?

Reply:
The `f (x := _) (y := _)` syntax is very general, not just for projections. When working with functions/methods with many implicit arguments, it becomes impractical to use the `@` syntax for instance. For example, if I had `Arguments some_big_lemma {arg1 arg2 arg3} arg4`, it may always be that case that when arg4 is supplied, the implicit ones can be inferred by Coq. However say I am in a proof and I don't quite know what arg4 is going to be. I might therefore write `some_big_lemma _`. Now Coq ends up putting all these arguments as existential variables when checking it. But perhaps I want to be a bit more restrictive, how can I supply a value for arg2 for instance?

The usual answer is to use the `@` syntax. This makes all implicit arguments explicit. So now I can write `@some_big_lemma _ value_for_arg2 _ _`. However now imagine there are more than 20 arguments, you can easily see how this becomes impractical.

The point of the `some_big_lemma (arg2 := value_for_arg2) _` syntax is to make exactly this problem manageable.